python/cudf/cudf/core/dataframe.py

```diff
@@ -401,57 +396,80 @@ class _DataFrameIlocIndexer(_DataFrameIndexer):
+            # appropriate indices.
+            (name,) = column_names
+            s = Series._from_data(
+                ca.__class__(
+                    {name: ca[name]},
+                    multiindex=ca.multiindex,
+                    level_names=ca.level_names,
+                ),
+                index=index,
+            )
+            return s._get(row_tag, row_key)
+        if column_names != list(self._frame._column_names):
+            frame = self._frame._from_data(
+                ca.__class__(
+                    {k: ca[k] for k in column_names},
+                    multiindex=ca.multiindex,
+                    level_names=ca.level_names,
+                ),
+                index=index,
+            )
         else:
-            if isinstance(arg[0], slice):
-                df = columns_df._slice(arg[0])
-            elif is_scalar(arg[0]):
-                index = arg[0]
-                if index < 0:
-                    index += len(columns_df)
-                df = columns_df._slice(slice(index, index + 1, 1))
-            else:
-                arg = (as_column(arg[0]), arg[1])
-                if is_bool_dtype(arg[0]):
-                    df = columns_df._apply_boolean_mask(arg[0])
-                else:
-                    df = columns_df._gather(arg[0])
-
-        # Iloc Step 3:
-        # Reindex
-        if df.shape[0] == 1:  # we have a single row without an index
-            df.index = as_index(self._frame.index[arg[0]])
-
-        # Iloc Step 4:
-        # Downcast
-        if self._can_downcast_to_series(df, arg):
-            return self._downcast_to_series(df, arg)
-
-        if df.shape[0] == 0 and df.shape[1] == 0 and isinstance(arg[0], slice):
-            df._index = as_index(self._frame.index[arg[0]])
-        return df
+            frame = self._frame
+        if row_tag is indexing_utils.IndexTag.MAP:
+            return frame._gather(
+                row_key,
+                keep_index=True,
+                nullify=False,
+                normalize_and_check=False,
+            )
+        elif row_tag is indexing_utils.IndexTag.MASK:
+            return frame._apply_boolean_mask(
+                row_key, keep_index=True, normalize_and_check=False
+            )
+        elif row_tag is indexing_utils.IndexTag.SLICE:
+            return frame._slice(cast(slice, row_key))
+        elif row_tag is indexing_utils.IndexTag.SCALAR:
+            result = frame._gather(
+                row_key,
+                keep_index=True,
+                nullify=False,
+                normalize_and_check=False,
+            )
+            # Attempt to turn into series.
+            try:
+                # Behaviour difference from pandas, which will merrily
+                # turn any heterogeneous set of columns into a series if
+                # you only ask for one row.
+                new_name = result.index[0]
+                result = Series._concat(
```

Comment:
No need to check if we can downcast to a series, we know we must try here because the indexing row tag was `SCALAR`.